src/format_string.h

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#include <ostream>
+#include <regex>
+#include <vector>
+
+#include "printf.h"
+
+namespace bpftrace {
+
+/**
+ * validate_fmt makes sure that the type are valid for the format specifiers
+ */
+std::string validate_format_string(const std::string &fmt,
+                                   std::vector<Field> args);
+
+struct Field;
+/*
+**
+*/
+class FormatString
+{
+private:
+  /**
+   * Split the format string on format specifiers, e.g.
+   * 'foo %s bar' -> [ 'foo %s', 'bar' ]
+   */
+  void split();
+
+public:
+  FormatString() = default;
```

Comment:
shouldn't need this but got some confusing template errors. Will try to remove it

Reply:
Don't see a way to avoid this, unless we change the `vector<tuple<FormatString, vector<Fields>>` into something with a custom serializer that does it correctly. Unless I'm doing something wrong.

Having this also keeps the serializer for this class easy.